src/NUnitFramework/framework/Constraints/DictionaryContainsKeyConstraint.cs

```diff
@@ -65,32 +71,138 @@ public override string Description
             Func<TMemberType, TCollectionType, bool> invertedComparison = (actual, expected) => comparison.Invoke(expected, actual);
 
+            _deprecatedFlag = true;
             base.Using(EqualityAdapter.For(invertedComparison));
             return this;
         }
+
+        /// <summary>
+        /// Flag the constraint to use the supplied Comparison object.
+        /// </summary>
+        /// <param name="comparer">The IComparer object to use.</param>
+        /// <returns>Self.</returns>
+        [Obsolete("Deprecated, use Does.ContainKey")]
+        public new CollectionItemsEqualConstraint Using<T>(Comparison<T> comparer)
+        {
+            _deprecatedFlag = true;
+            return base.Using(comparer);
+        }
+
+        /// <summary>
+        /// Checks if the key is contained in a "keyed item container".
+        /// </summary>
+        /// <param name="actual">Keyed container.</param>
+        /// <returns>method to call.</returns>
+        private MethodInfo GetContainsKeyMethod(object actual)
+        {
+            if (actual == null) throw new ArgumentNullException(nameof(actual));
+            var instanceType = actual.GetType();
+
+            var method = FindContainsKeyMethod(instanceType)
+                         ?? instanceType
+                            .GetInterfaces()
+                            .Concat(GetBaseTypes(instanceType))
+                            .Select(FindContainsKeyMethod)
+                            .FirstOrDefault(m => m != null);
+
+            return method;
+        }
+
+        /// <summary>
+        /// Looks for a base type that implements ContainsKey method
+        /// </summary>
+        /// <param name="type">Type to look for method ContainsKey</param>
+        /// <returns>Returns the method to call.</returns>
+        private MethodInfo FindContainsKeyMethod(Type type)
+        {
+            var methods = type.GetMethods(BindingFlags.Instance | BindingFlags.Public);
+            var method = methods.FirstOrDefault(m =>
+                m.ReturnType == typeof(bool)
+                && m.Name == "ContainsKey"
+                && m.GetParameters().Length == 1);
+
+            if (method == null && type.GetTypeInfo().IsGenericType)
+            {
+                var definition = type.GetGenericTypeDefinition();
+                var tKeyGenericArg = definition.GetGenericArguments().FirstOrDefault(typeArg => typeArg.Name == "TKey");
+
+                if (tKeyGenericArg != null)
+                {
+                    method = definition
+                             .GetMethods(BindingFlags.Instance | BindingFlags.Public)
+                             .FirstOrDefault(m =>
+                                 m.ReturnType == typeof(bool)
+                                 && m.Name == "Contains"
+                                 && m.GetParameters().Any()
+                                 && m.GetParameters().First()?.ParameterType == tKeyGenericArg);
```

Comment:
`m.GetParameters().First()` can't return null, so `?.` is out of place there. However, these two lines could be merged to `m.GetParameters().FirstOrDefault()?.ParameterType ==`.